LogicLayer/Friends/FriendsService.cs

```diff
@@ -0,0 +1,133 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data.SqlClient;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading.Tasks;
+using GameBoard.DataLayer.Entities;
+using GameBoard.DataLayer.Enums;
+using GameBoard.LogicLayer.Friends.Dtos;
+using GameBoard.LogicLayer.UserSearch.Dtos;
+using GameBoard.DataLayer.Repositories;
+using GameBoard.LogicLayer.Friends.Exceptions;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Query.Expressions;
+using Microsoft.Extensions.Logging;
+using Remotion.Linq.Parsing;
+
+namespace GameBoard.LogicLayer.Friends
+{
+    internal sealed class FriendsService : IFriendsService
+    {
+        private readonly IGameBoardRepository _repository;
+
+        public FriendsService(IGameBoardRepository repository)
+        {
+            _repository = repository;
+        }
+
+        public async Task<IEnumerable<UserDto>> GetFriendsByUserNameAsync(string userName)
+        {
+            string normalizedUserName = userName.ToUpper();
+            var user = _repository.ApplicationUsers.Where(u => u.NormalizedUserName == normalizedUserName);
+
+            var friendsInvitedByMe = user
+                .Include(u => u.SentRequests)
+                .SelectMany(u => u.SentRequests)
+                .Where(f => f.FriendshipStatus == FriendshipStatus.Lasts)
+                .Include(f => f.RequestedTo)
+                .Select(f => new UserDto(f.RequestedTo.Id, f.RequestedTo.UserName, f.RequestedTo.Email));
+
+            var friendsThatInvitedMe = user
+                .Include(u => u.ReceivedRequests)
+                .SelectMany(u => u.ReceivedRequests)
+                .Where(f => f.FriendshipStatus == FriendshipStatus.Lasts)
+                .Include(f => f.RequestedBy)
+                .Select(f => new UserDto(f.RequestedBy.Id, f.RequestedBy.UserName, f.RequestedBy.Email));
+
+            var allFriends = await friendsInvitedByMe // can I avoid using await here?
+                .Union(friendsThatInvitedMe)
+                .OrderBy(u => u.UserName) // maybe order by NormalizedUserName? However, I cannot do that with UserDto.
```

Comment:
Maybe do, maybe don't - decide and remove the comment.

Reply:
I decided to order by normalized user name, because it makes more sense to me (and with ApplicationUser -> UserDto conversion logic after toListAsync it is accually possible :) )